File: polars-rust/src/packs/completeness.rs
use polars::prelude::{col, JoinArgs, JoinType, LazyFrame};

/// Run the Completeness Report
pub fn completeness_test(gl: &LazyFrame, tb: &LazyFrame) -> LazyFrame {
    // Summarize TB by account
    let tb_groupby_columns = [
        col("TB_Fiscal_Year"),
        col("TB_Business_Unit_Code"),
        col("TB_Account_Number"),
    ];
    let tb_total_by_account = tb.clone().group_by(&tb_groupby_columns).agg([
        //TODO Rounding kills streaming
        //col("TB_Amount_Opening_Balance").sum().round(2),
        //col("TB_Amount_Ending_Balance").sum().round(2),
        col("TB_Amount_Opening_Balance").sum(),
        col("TB_Amount_Ending_Balance").sum(),
    ]);

    // Summarize GL by account
    let gl_groupby_columns = [
        col("GL_Fiscal_Year"),
        col("GL_Business_Unit_Code"),
        col("GL_Account_Number"),
    ];
    let gl_total_by_account = gl
        .clone()
        .group_by(&gl_groupby_columns)
        //TODO Rounding kills streaming
        //.agg([col("GL_Local_Amount").sum().round(2)]);
        .agg([col("GL_Local_Amount").sum()]);

    // Combine TB and GL Data
    let gl_tb_total_by_account = tb_total_by_account.join(
        gl_total_by_account,
        &tb_groupby_columns,
        &gl_groupby_columns,
        //TODO Rounding kills streaming, but so do Outer Joins, so just got to live with this one
        //JoinArgs::new(JoinType::Outer { coalesce: true }),
        JoinArgs::new(JoinType::Left),
    );

    // Add the difference between the GL and TB
    let gl_tb_total_by_account = gl_tb_total_by_account.with_column(
        (col("TB_Amount_Opening_Balance").fill_null(0_f64)
            + col("GL_Local_Amount").fill_null(0_f64)
            - col("TB_Amount_Ending_Balance").fill_null(0_f64))
        //TODO Rounding kills streaming
        //.round(2)
        .alias("Difference"),
    );

    // Filer for just columns needed, rename them, and sort to make compressions nice
    gl_tb_total_by_account
        .select([
            col("TB_Business_Unit_Code").alias("Business_Unit_Code"),
            col("TB_Fiscal_Year").alias("Fiscal_Year"),
            col("TB_Account_Number").alias("Account_Number"),
            col("TB_Amount_Opening_Balance").alias("Opening_Balance"),
            col("GL_Local_Amount").alias("Activity"),
            col("TB_Amount_Ending_Balance").alias("Ending_Balance"),
            col("Difference"),
        ])
        .sort_by_exprs(
            [
                col("Business_Unit_Code"),
                col("Fiscal_Year"),
                col("Account_Number"),
            ],
            [false; 3],
            false,
            false,
        )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::io::{read_gl, read_tb};
    use std::path::Path;

    #[test]
    fn does_completeness_run() {
        let gl = read_gl(Path::new("../gitignore/data/gl.parquet/gl_0000.parquet")).unwrap();
        let tb = read_tb(Path::new("../gitignore/data/tb.parquet/tb_0000.parquet")).unwrap();

        let unbalanced: LazyFrame = completeness_test(&gl, &tb);
        let _ = unbalanced.collect().unwrap();
    }
}

File: pyspark/src/packs/completeness.py
from pyspark.sql.dataframe import DataFrame  # type: ignore
from pyspark.sql.functions import sum  # type: ignore


# Compute the Unbalanced JE Report
def completeness_test(gl: DataFrame, tb: DataFrame) -> DataFrame:

    # Summaraize GL by account number
    gl_groupby_columns = [
        "GL_Fiscal_Year",
        "GL_Business_Unit_Code",
        "GL_Account_Number",
    ]
    gl_by_account = gl.groupby(gl_groupby_columns).agg(
        sum("GL_Local_Amount").alias("GL_Local_Amount")
    )

    # Summaraize TB by account number
    tb_groupby_columns = [
        "TB_Fiscal_Year",
        "TB_Business_Unit_Code",
        "TB_Account_Number",
    ]
    tb_by_account = tb.groupby(tb_groupby_columns).agg(
        sum("TB_Amount_Opening_Balance").alias("TB_Amount_Opening_Balance"),
        sum("TB_Amount_Ending_Balance").alias("TB_Amount_Ending_Balance"),
    )

    # Combine GL and TB data
    gl_tb_by_account = gl_by_account.join(
        tb_by_account,
        on=[
            gl_by_account["GL_Fiscal_Year"] == tb_by_account["TB_Fiscal_Year"],
            gl_by_account["GL_Business_Unit_Code"]
            == tb_by_account["TB_Business_Unit_Code"],
            gl_by_account["GL_Account_Number"] == tb_by_account["TB_Account_Number"],
        ],
        how="outer",
    )

    # Fill in Null columns from Outer Join
    gl_tb_by_account = gl_tb_by_account.na.fill(0)

    # Compute any completness differances
    gl_tb_by_account = gl_tb_by_account.withColumn(
        "Differance",
        (
            gl_tb_by_account["TB_Amount_Opening_Balance"]
            + gl_tb_by_account["GL_Local_Amount"]
            - gl_tb_by_account["TB_Amount_Ending_Balance"]
        ),
    )

    # Filter just for the columns we want and cleanup column names
    column_list = [
        "TB_Fiscal_Year",
        "TB_Business_Unit_Code",
        "TB_Account_Number",
        "TB_Amount_Opening_Balance",
        "GL_Local_Amount",
        "TB_Amount_Ending_Balance",
        "Differance",
    ]
    completness_report = (
        gl_tb_by_account[column_list]
        .withColumnRenamed("TB_Fiscal_Year", "Fiscal_Year")
        .withColumnRenamed("TB_Business_Unit_Code", "Business_Unit_Code")
        .withColumnRenamed("TB_Account_Number", "Account_Number")
        .withColumnRenamed("TB_Amount_Opening_Balance", "Opening_Balance")
        .withColumnRenamed("GL_Local_Amount", "Activity")
        .withColumnRenamed("TB_Amount_Ending_Balance", "Ending_Balance")
    )
    return completness_report
